pkg/operator/operator_test.go

```diff
@@ -0,0 +1,197 @@
+			csr, gerr = apiV1.Get(ctx, fqdn, metav1.GetOptions{})
+			if apierrors.IsNotFound(gerr) {
+				return false, nil
+			}
+			// Check if certificate has been approved by the signing function.
+			if len(csr.Status.Conditions) < 1 {
+				return false, nil
+			} else {
+				return csr.Status.Conditions[0].Type == v1.CertificateApproved, nil
+			}
+		}); err != nil {
+			t.Errorf("timeout waiting for CSR: %s", err)
+		}
+
+		// Mock the kube-apiserver issuing the certificate by writing toy data to status.
+		// This will unblock the keypair generation goroutine.
+		csr.Status.Certificate = certBytes
+		if _, err := apiV1.UpdateStatus(ctx, csr, metav1.UpdateOptions{}); err != nil {
+			t.Errorf("updating csr while issuing cert: %s", err)
+		}
+	}()
+}
+
+func readKeyAndCertFiles(dir string, t *testing.T) ([]byte, []byte) {
+	outCert, err := ioutil.ReadFile(path.Join(dir, "tls.crt"))
+	if err != nil {
+		t.Fatalf("error reading from cert file: %v", err)
+	}
+	outKey, err := ioutil.ReadFile(path.Join(dir, "tls.key"))
+	if err != nil {
+		t.Fatalf("error reading from key file: %v", err)
+	}
+	return outCert, outKey
+}
+
+func TestEnsureCertsExplicit(t *testing.T) {
+	dir, err := ioutil.TempDir("", "test_ensure_certs")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	for _, tc := range []struct {
+		desc       string
+		opts       Options
+		expectCert string
+		expectKey  string
+		expectErr  bool
+	}{
+		{
+			desc:       "input key/cert",
+			opts:       Options{Key: "a2V5", Cert: "Y2VydA=="},
+			expectCert: "cert",
+			expectKey:  "key",
+			expectErr:  false,
+		},
+		{
+			desc:      "bad cert",
+			opts:      Options{Cert: "not a cert", Key: "not a key"},
+			expectErr: true,
+		},
+		{
+			desc:      "cert and no key",
+			opts:      Options{Cert: "cert", Key: ""},
+			expectErr: true,
+		},
+		{
+			desc:      "no cert and key",
+			opts:      Options{Cert: "", Key: "key"},
+			expectErr: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			op := Operator{
+				opts: tc.opts,
+			}
+			_, err := op.ensureCerts(context.Background(), dir)
+			if err != nil {
+				if !tc.expectErr {
+					t.Fatalf("want err: %v; got %v", tc.expectErr, err)
+				}
+				return
+			}
+			// Test outputed files.
+			outCert, outKey := readKeyAndCertFiles(dir, t)
+			if string(outCert) != tc.expectCert {
+				t.Errorf("want cert %v; got %v", tc.opts.Cert, string(outCert))
+			}
+			if string(outKey) != tc.expectKey {
+				t.Errorf("want key %v; got %v", tc.opts.Key, string(outKey))
+			}
+		})
+	}
+}
+
+func TestEnsureCertsServerSigned(t *testing.T) {
+	var (
+		timeout     = 3 * time.Second
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+		certBytes   = []byte{1, 2, 3, 4}
+	)
+	t.Cleanup(cancel)
+
+	dir, err := ioutil.TempDir("", "test_ensure_certs")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	client := fake.NewSimpleClientset()
+	mockKubeAPIServer(ctx, client, t)
```

Comment:
Can we switch the go routine such that you issue it here instead, while changing `mockKubeAPIServer` to be a blocking function?
i.e.
`go mockKubeAPIServer(ctx, client, t)`

and remove the go routine in the actual function. That reads a little cleaner to me and I can tell what's going on better.